records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+# Database performance issues
+
+Date: 2021-07-16
+
+## Status
+
+**DRAFT**
+
+## Context
+
+Because Notify suffers from regular daily incident that originates from its
+database when clients blasts notification, we need to review some design aspect
+of the stack and bring more control to specific elements.
+
+In the past months, the current architecture, environment and configuration have
+been tweaked to achieve a certain balance that could take on the expected load.
+To list a few of these tweaks:
+
+* the number of maximum pool connection in individual application instance,
+* the AWS instance classes,
+* the resources requests and limits of Kubernetes containers,
+* the horizontal pod autoscaling.
+
+These serve well the application but there were already signs of the
+application's limit: we rarely if ever meet our SLAs and we had regular daily
+incidents that foretold of the difficulties ahead.
+
+This document propose various options to address thorny issues that our
+database is having. [Click here for schema of our database](diagrams/2021-07-16.scalability.database-bottleneck/relationships.real.compact.png)
+at the time of this writing.
+
+## Timeline of an incident
+
+1. A blast of notifications is sent to Notify through the API.
+2. As the notification-api component saves the notifications to be sent in
+   the database, it commits each of these individually[1].
+3. The notification is then sent to the appropriate queue, depending on its
+   type, scheduling and used key.
```

Comment:
There is a missing step here. Once the notification goes to the queue the status of the Notification is updated from created to sending. This is important because its an update operation ( under load can trigger autovac )

Reply:
Hmm this is covered by step #5 but maybe the chronology isn't clear here. Let me rework this.